design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,259 @@
+  def set_virtual_ref(
+      self,
+      key: str,
+      location: str,
+      offset: int,
+      length: int,
+      checksum: ETag | LastModified | None = None,
+      validate_containers: bool = False
+  ) -> None: ...
+```
+
+`checksum` can be set to the in-object-store ETag of the file containing the virtual chunk, or to its last modified time. In the case of last modified time, the value
+doesn't need to be precise. The system will only fail to provide the chunk if the object's last-modified-at is larger than the passed value. So, passing `checksum = LastModified(datetime.now())` will usually be enough.
+
+This new argument will help us satisfy requirement 4.
+
+The interpretation of the `location` argument also changes. `location` must be an absolute url that includes a protocol. The protocol can be:
+
+* s3://
+* gcs://
+* tigris://  (is this the right protocol)
+* fs://  (is this right for local file system)
+* etc. for other supported object stores
+* arbitrary-virtual-chunk-container-name://  (example: models://)
+
+Depending on the protocol used, Icechunk will use a different mechanism to resolve the chunk, as explained below.
+
+If `validate_containers` is set to `True`, the function will fail if the protocol passed is not registered in a virtual chunk container. Using `validate_containers=False` is useful, if containers will be declared after the write.
+
+### Reading virtual chunks
+
+When a virtual chunk is requested, Icechunk will find in the `location` field of the stored reference, an absolute URL with a specific protocol. It will need to map this location to a given object store, bucket, path and set of credentials.
+
+The mapping is done by comparing the protocol and path to the set of configured virtual chunk containers, plus a set of default virtual chunk containers.
+
+This is how the mapping from location to cloud+bucket+prefix+credentials works:
+
+* Using the `location` protocol we find all containers that match the protocol
+* We sort the matching containers in order of descending prefix length
+* We find the first one that has a `bucket_prefix` that is a prefix of the `location` path (this will be the most specific container that matches)
+* If no containers match, the chunk request cannot be fulfilled and we fail with an error message indicating the missing config.
+* We use the rest of the data in the matched container to instantiate an object store client (or retrieve it from cache)
+* Credentials for the client must be passed by the user when the repository is instantiated, by indicating the container name to which they apply
+
+We have one default container for each supported object store, the container has protocol and name s3, gcs, tigris, etc. and has an empty `bucket_prefix` and no region configured. These default containers, just like any others set in the persistent repo configuration, can be overridden when the repository is instantiated.
```

Comment:
So, to be precise, you are envisioning a structure like:
```json
[{"name": "acontainer",
   "prefix": "gcs://bucket/path/prefix",
   "options": {"region": "US-EAST1"}},
 {...}
]
```
where the "name" handle is what the user passes to change the default?

and "prefix" is literal, it matches any paths that start with that string.